waspc/examples/todoApp/migrations/prisma.schema-checksum

```diff
@@ -0,0 +1 @@
+00a14d3339ef0e8cb44b04752e59656d41189df28b6ac02e8c574f695a047bbc
```

Comment:
Adding our checksum file for the sample app. In real apps it will be in their migrations dir as well so we can even warn them on a fresh checkout, before they do `wasp db migrate-dev`, if they change their entity info.

Reply:
Arg, yeah I think you are right. There was a different error that masked this https://github.com/wasp-lang/wasp/issues/427 on first run which was normally solved by a `wasp db migrate-dev`, but we do need on a fresh checkout to handle that empty db case. I may need to actually draw out a state diagram for all the cases since it's getting harder and harder to wrap my head around all the permutations of source states vs db states 🤯

Reply:
Looking back through the discussion, the use case Martin and I were hoping to solve here was if they migrated, and either did a fresh checkout or did a clean, to not warn them unnecessarily they need to migrate again. However, with this file in the source tree I don't think we can solve for the actual problem if they need to migrate anymore. I believe I will have to remove this file from the repo and we may just have to warn them on checkout or clean they need to migrate (even if redundant), and this file being present will serve as "proof" they migrated locally. I will confirm with the table though. Sorry for missing that and good catch 👍🏻

Reply:
Ok, in thinking it through a bit more, I've come up with the following:
- We cannot include the checksum in the repo, as then we have no way to determine if they should migrate on a fresh checkout and empty db. We need to use the checksum as local "proof" they ran `wasp db migrate-dev`, so it should remain in the generated project.
- We cannot, however, warn them if there is no checksum file either, as on a new project or those without any entities, they should not/cannot migrate anyways, so they cannot generate a checksum.
- We can only start warning after they have successfully executed `wasp db migrate-dev` locally at least once (and thus created a checksum in the generated project).
  - This means though that we are missing several cases, including:
    - Need to run migrate on fresh checkout and empty db
    - Making changes to prisma.schema before checksum is generated

Therefore, the only check we can do is if they have a checksum and it differs from the current schema, we warn. Or if they have a checksum and no schema they are in some crazy state and warn to clean. Still an improvement but I cannot see how we can detect any other nice to have cases.

Thankfully, this is how the checking logic already works now, so after removing checksum from migrations/repo it should behave as above. Does this reflect your thinking @Martinsos @matijaSos? Am I missing anything else? Thanks

Reply:
I think I would recommend we do in fact remove the checksum from the repo, even it won't help us solve for the fresh checkout and empty db case, to cement the fact that it is a local artifact to help with local migrate decision making.

Reply:
Great job @shayneczyzewski and @matijaSos  for catching this one, I also totally missed it!

Ok, so I read through what you wrote @shayneczyzewski and I also tried reasoning about it on my own.

I agree that we shouldn't store the checksum in the source project -> while the checksum does tell us that those migrations were generated with the specific prisma schema, we don't know if they were applied on the local database, they might have been applied to a local database by developer A and then checked-out by developer B, and therefore developer B does need to run `wasp db migrate-dev`. However, we can store the checksum in the generated project, therefore locally, in order to track for which version of prisma.schema `wasp db migrate-dev` was called last time locally.

In the case of `wasp clean` or new checkout, there is no generated project, therefore no migrations dir in the generated project and no checksum in the generated project. In that case, why not warn user that they should do migrate-dev? You said because we don't know if they have any entities or not -> if they don't they shouldn't have to run migrate-dev. That makes sense -> but I guess we can check if they have any entities or not and act accordingly, can't we? So if they have any entities, we could warn them that they should run `wasp db migrate-dev`, if migrations directory or checksum file is missing, but if they don't we wouldn't warn them.

Regarding `prisma generate` -> hm this is interesting. So `prisma generate` generates JS/TS bindings for prisma, maybe also something else like parts or whole CLI/Studio and similar. Anyway, it generates prisma artifacts/code, all based on the schema.prisma.
Now, it seems to me that with the warning system we described above, we are covered -> we are warning them to run `wasp db migrate-dev`, which will also do generate, so that is good, all situations are covered -> if they change schema.prisma, we suggest they run it, and if we are not sure if they run it (missing checksum), we again ask them to run it. Except if there are no entities, then we don't ask them anything which is also ok.

However, there is one thought I had -> during development, you might be making changes to schema.prisma -> adding lines, models, ... . Also, modifying some other code based on that, to keep up with the changes. You might not want to run the `migrate` immediately, since you don't want to create a ton of migrations as you are working on it, you first want to get it to compile correctly (if you are using TypeScript) and pass some unit tests. In sucn case, we could be extra smart and run `prisma generate` on every change to schema.prisma automatically! `migrate` we don't want to run automatically, because we don't want to presume when they are ready to do a migration, as migrations introduce some complexity and can go wrong and require some manual intervention and similar. But, `generate` is just generating code -> nothing to go wrong there, and no reason why they wouldn't want that to happen instantly on saving of schema.prisma, right? So we could make the running of `prisma generate` a part of the Wasp.Db.Generator -> when ever we are compiling Wasp code, we would also call prisma generate, if schema.prisma changed since the last compile or is missing (and all of this we would do only if there are entities declared in Wasp).

This said, the last thing with `prisma generate` is an improvement, so we could leave that for a separate issue and for now go with just `wasp db migrate-dev`.

Does this make sense, did I miss something? I tried thinking through all the cases but I might have missed something again.

Reply:
Hey @Martinsos - thanks for the reply!

Cool, sounds we are in agreement to make it local. Per the warning on fresh checkout/clean:

> That makes sense -> but I guess we can check if they have any entities or not and act accordingly, can't we? So if they have any entities, we could warn them that they should run wasp db migrate-dev, if migrations directory or checksum file is missing, but if they don't we wouldn't warn them.

I agree this would be nice. The only way I can think atm to use as a proxy for "if they have entities" is the existence of the migrations dir. I think that should be a good enough thing to check though. So in this case it will be "if a wasp migrations dir exists, but not a generated project checksum, warn they should migrate". I think that will work 👍🏻 

> But, generate is just generating code -> nothing to go wrong there, and no reason why they wouldn't want that to happen instantly on saving of schema.prisma, right? So we could make the running of prisma generate a part of the Wasp.Db.Generator -> when ever we are compiling Wasp code, we would also call prisma generate, if schema.prisma changed since the last compile or is missing (and all of this we would do only if there are entities declared in Wasp).

I will think on that more. Worst case we can punt it into the new issue I created and do that shortly after this one, but if small and aligned I may be able to sneak it in.

Thanks all!

Reply:
Can't we read from AppSpec if they have entities?

Reply:
Unfortunately, I do not think we will have access to an AppSpec at the time we need to do the check. This is currently being called in two places:

1) in CLI watch
```haskell
watch :: Path' Abs (Dir Common.WaspProjectDir) -> Path' Abs (Dir Wasp.Lib.ProjectRootDir) -> IO ()
```

and 2) in Generator start, which is called by CLI start:
```haskell
-- Generator
start :: Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
```

All the checks have to take place outside of the main Generation/compile step since Generation happens before every migrate, so I could not make it a precondition. So I am making due with only `Path' Abs (Dir ProjectRootDir)`.

Reply:
Hmmmm ok interesting!

I will try to recap things a bit for my sake:
1. We want to warn use that they need to run `wasp db migrate-dev`.
2. We do that by checking if schema.prisma in the generated code matches the last schema that we used for performing migration -> if not, it means this new schema might need migration, so we show that warning.
3. We store checksum in the generated dir, so it is local, which makes sense.
4. If there is no checksum, we go safe route and show potentially redundant warning that they need to migrate.
5. There is an exception: if user has no entities, then we should not show any warning.

When looking at this, I imagine we could get it working by throwing a warning (not a fatal error) somewhere from inside the compiler -> it could be right after the generation step of compiler but still inside of compiler, or it could be from inside the generator. This way we would have access to AppSpec, and since it is a warning, it won't stop the execution. The question is how do we make it be just a warning though, but the truth is we will likely have other warnings also in the future, so we will probably need to implement support for them anyway at some point.

I am talking a bit abstract while you are actually on the "terrain", so maybe I am missing some practical issues here -> do you think this would be doable, how hard is it?
Could you add additional step into Generator, somewhere in Generator.hs maybe, that checks for this and returns warning, something in that direction, would that work ok?

From what I got you didn't want to make this part of compilation because it would crash the compilation and therefore also prevent us from doing migrate since it calls compilation first. But if it is a warning and doesn't crash anything, then that would not be a problem right? But this makes it impossible for you to access AppSpec, which means we can't make "smart" decisions, like checking there are any entities present in the project.

Reply:
Yep, your recap sounds accurate to me! I agree, the non-fatal warning approach during compilation/generation is ideal. It makes me think pausing on this until generator errors https://github.com/wasp-lang/wasp/issues/123 is done feels like the best strategy. Otherwise, we will likely have to refator this after anyways.